### DISCUSSION TITLE
using grove i2c motor drivers

### DISCUSSION THREAD

**Discussion by davidf-2013** [fjp/2wd-robot#46] [General]
Hi!  First off, thanks for putting this stuff out there, its great.  Here is my issue:  it seems that things have migrated to the adafruit motor driver, but I am using the grove i2c motor driver..
 
from diffbot_bringup/launch/minimal.launch:
Deprecated: Use when having a grove_motor_driver together with the grove_motor_driver ROS package
&lt;!--node name="motor_driver" pkg="grove_motor_driver" type="motor_driver.py" respawn="false" output="screen" ns="diffbot"/--&gt;

I can't seem to find the sources for this motor_driver.py or the grove_motor_driver ROS package.. any hints?
Thanks in advance!

**Comment by fjp** [comment]
Hi @davidf-2013 sorry for the confusion here. You can find this package still here https://github.com/ros-mobile-robots/grove_motor_driver
A short note about this package: It is wrapper around https://github.com/DexterInd/GrovePi/tree/master/Software/Python/grove_i2c_motor_driver and it is based on python. This is working fine but might not give you the best performance.

An alternative would be to use https://github.com/Seeed-Studio/Grove_I2C_Motor_Driver_v1_3 as basis and add a new ros wrapper package around it.

But take care which PID controller you are going to use. In case you want to use the grove_motor_driver package you need to go with the high-level PID controller (running in the hardware_interface). This will publish two motor velocities, see the following

https://github.com/ros-mobile-robots/diffbot/blob/327f39559a7d7f2a49ea5a9534b941f7777e483b/diffbot_base/src/diffbot_hw_interface.cpp#L58-L60

and 

https://github.com/ros-mobile-robots/diffbot/blob/327f39559a7d7f2a49ea5a9534b941f7777e483b/diffbot_base/src/diffbot_hw_interface.cpp#L188-L227

The recommended way would be to implement the grove motor drive in the low level firmware running on the microcontroller (e.g. Teensy board), This way you would use the PID controllers running on the Teensy firmware. Let me know if you need help with that or more guidance on how to use the grove_motor_driver package together with the high level PID controllers.

**Reply by davidf-2013** [reply]
thanks for the reply and the information!  i'll probably try to just use the python wrappers/pi based pid first, and then i do think i can implement the teensy bit later (or see if its already there on remo side?)..  it doesn't seem horrific but it would be "yet another thing" to make/maintain the extra node and i am trying to replicate the work according to your tutorials.. thanks agian!

**Reply by fjp** [reply]
You are welcome!

Sure, starting with the high level PID makes sense and I can support in doing that, just let me know if you need help. 

For Remo and the low level PID currently there is only support for the [DC Motor + Stepper FeatherWing](https://www.adafruit.com/product/2927) through the [`motor_controller_interface`](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/base_controller/lib/motor_controller_interface/motor_controller_interface.h). In case you decide to go that route, running the low level PID together with the grove motor driver board, you could make use of that interface too. This interface should make the integration of new motor drivers quite easy. Adding support for the grove motor driver with its https://github.com/Seeed-Studio/Grove_I2C_Motor_Driver_v1_3 library is also on my todo list but there are other tasks which have higher priority (such as the IMU integration). However, if you decide to try the low level PID together with the grove motor driver, a PR would be most welcome :-)
